src/main/java/com/microsoft/sqlserver/jdbc/ParameterMetaDataCache.java

```diff
@@ -240,4 +243,68 @@ private static AbstractMap.SimpleEntry<String, String> getCacheLookupKeys(SQLSer
 
         return new AbstractMap.SimpleEntry<>(cacheLookupKey, enclaveLookupKey);
     }
+
+    /**
+     * Creates a copy of the enclave keys to be sent to the crypto cache. This allows reconnection when table
+     * information has changed but session, and enclave information, is the same.
+     * 
+     * @param keysToBeCached
+     *        The keys sent to the cryptocache
+     * @return A copy of the enclave keys, stored in the cryptoCache
+     */
+    private Map<Integer, CekTableEntry> copyEnclaveKeys(Map<Integer, CekTableEntry> keysToBeCached) {
+        Map<Integer, CekTableEntry> enclaveKeys = new ConcurrentHashMap<>();
+        for (Map.Entry<Integer, CekTableEntry> entry : keysToBeCached.entrySet()) {
+            int ordinal = entry.getKey();
+            CekTableEntry original = entry.getValue();
+            CekTableEntry copy = new CekTableEntry(ordinal);
+            for (EncryptionKeyInfo cekInfo : original.columnEncryptionKeyValues) {
+                copy.add(cekInfo.encryptedKey, cekInfo.databaseId, cekInfo.cekId, cekInfo.cekVersion,
+                        cekInfo.cekMdVersion, cekInfo.keyPath, cekInfo.keyStoreName, cekInfo.algorithmName);
+            }
+            enclaveKeys.put(ordinal, copy);
+        }
+        return enclaveKeys;
+    }
+}
+
+
+/**
+ * Represents a cache of all queries for a given enclave session.
+ */
+class CryptoCache {
+    /**
+     * The cryptocache stores both result sets returned from sp_describe_parameter_encryption calls. Column Encryption
+     * Key data in cekMap, and parameter data in paramMap.
+     */
+    private final ConcurrentHashMap<String, Map<Integer, CekTableEntry>> cekMap = new ConcurrentHashMap<>(16);
+    private ConcurrentHashMap<String, ConcurrentHashMap<String, CryptoMetadata>> paramMap = new ConcurrentHashMap<>(16);
+
+    ConcurrentHashMap<String, ConcurrentHashMap<String, CryptoMetadata>> getParamMap() {
+        return paramMap;
+    }
+
+    void replaceParamMap(ConcurrentHashMap<String, ConcurrentHashMap<String, CryptoMetadata>> newMap) {

```

Comment:
Just a question related to the cache trimming. When we trim the cache, how do we know we are removing stale entries?

Reply:
We're not exactly removing stale entries, just the ~300 oldest, which will have occurred 2000 queries ago. If 2000 queries have been executed since the first one, I think it would be safe to assume that first one is no longer cached. I can't think of an alternative, only expanding the cache size, but that would just have memory concerns.

Reply:
Hmm, I'm little worried about entry trim then, because we remove entries by iterating from the first entry. Maps do not preserve order of insertion.

Reply:
Good catch. We should have used ConcurrentLinkedHashMap in the original implementation to ensure order is preserved.

Reply:
Yup thanks for that, and actually, that exposed a possibility I didn't test for. I added a test for trimming which should have been there in the first place.